### PR TITLE
MAINT: refactor: arraydescr_new now does not call PyArray_DescrNew 

### PR DESCRIPTION
**Pull request by mattip**
I started making dtype classes so that the `type(np.dtype('int32'))` is no longer `numpy.dtype` rather a new `numpy.int_type` type that inherits from `numpy.dtype`, the first step in the [plan](https://hackmd.io/cVdS9UyBRayZF-tIW1lC0g#) to refactor dtypes. Rather than a single large PR, I will try to do this in a set of smaller PRs, with an "opt-out" `#ifdef` ... `#else` ... `#endif` around the code.

This PR is the first in the series - a refactoring of the dependency between `arraydescr_new` and `PyArray_DescrNew` so that the latter now calls the former instead of the other way around. ~I also refactored the function to create `datetime64` metadata and moved it to a different place in the file.~

Benchmarking seemed to show no real difference in performance, although there is alot of variance in the `bench_core.CountNonzero.time_count_nonzero` tests

### CODE REVIEW

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 1]
```diff
@@ -4166,6 +4166,55 @@ small_correlate(const char * d_, npy_intp dstride,
...
+}
+
+/*
+ * Initializes the c_metadata field
+ * must not be static, gcc 4.1.2 on redhat 5 then miscompiles this function
+ * see gh-5163
+ *
+ */
+NPY_NO_EXPORT int
+_initialize_builtin_datetime_metadata(void)
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/multiarray/arraytypes.c.src]: refactored out of this PR

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 2]
```diff
@@ -4650,7 +4639,12 @@ set_typeinfo(PyObject *dict)
 
     /**end repeat**/
 
-    if (initialize_builtin_datetime_metadata() < 0) {
+    _builtin_descrs[NPY_DATETIME]->c_metadata = (NpyAuxData *)_initialize_builtin_datetime_metadata();
+    if (_builtin_descrs[NPY_DATETIME]->c_metadata == NULL) {
+        return -1;
+    }
+    _builtin_descrs[NPY_TIMEDELTA]->c_metadata = (NpyAuxData *)_initialize_builtin_datetime_metadata();
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/multiarray/arraytypes.c.src]: fixed and refactored out of this PR

File: numpy/core/src/multiarray/descriptor.c [hunk 3]
```diff
@@ -257,6 +257,9 @@ _convert_from_tuple(PyObject *obj, int align)
             return NULL;
         }
         PyArray_DESCR_REPLACE(type);
+        if (type == NULL) {
+            return NULL;
+        }
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/multiarray/descriptor.c]: As `PyArray_DESCR_REPLACE` calls `PyArray_DescrNew` which has various failure modes, it is a bugfix

File: numpy/core/src/multiarray/descriptor.c [hunk 4]
```diff
@@ -1698,10 +1707,22 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
 NPY_NO_EXPORT PyArray_Descr *
 PyArray_DescrNew(PyArray_Descr *base)
 {
-    PyArray_Descr *newdescr = PyObject_New(PyArray_Descr, &PyArrayDescr_Type);
+    PyObject *tuple = PyTuple_Pack(3,
+                                  /* dtype */ base,
+                                  /* align */ Py_False,
+                                  /* copy  */ Py_True);
+    PyObject *ret = PyObject_CallObject((PyObject*)Py_TYPE(base), tuple);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: `PyObject_CallFunctionObjArgs((PyObject *)Py_TYPE(base), base, Py_False, Py_True)` would be shorter here - or am I missing something?

File: numpy/core/src/multiarray/descriptor.c [hunk 5]
```diff
@@ -1733,7 +1754,7 @@ PyArray_DescrNew(PyArray_Descr *base)
         newdescr->subarray = PyArray_malloc(sizeof(PyArray_ArrayDescr));
         if (newdescr->subarray == NULL) {
             Py_DECREF(newdescr);
-            return (PyArray_Descr *)PyErr_NoMemory();
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: This `PyErr_NoMemory()` needs to remain

File: numpy/core/src/multiarray/descriptor.c [hunk 6]
```diff
@@ -1698,10 +1707,22 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
...
+                                  /* copy  */ Py_True);
+    PyObject *ret = PyObject_CallObject((PyObject*)Py_TYPE(base), tuple);
+    assert(ret != (PyObject*)base);
+    Py_DECREF(tuple);
+    return (PyArray_Descr*)ret;
+}
 
+/* Initialize the PyArray_Descr fields */
+static int
+_initialize_descr(PyArray_Descr *newdescr, PyArray_Descr *base)
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/multiarray/descriptor.c]: Adopting the second option, so we can, in the future, subclass dtype

File: numpy/core/src/multiarray/descriptor.c (hunk 4 above)
Comment (mattip) [reply numpy/core/src/multiarray/descriptor.c]: adopted

**Comment by eric-wieser** [comment]
> I started making dtype classes so that the type(np.dtype('int32')) is no longer numpy.dtype rather a new numpy.int_type type that inherits from numpy.dtype,

I'm not sure this is where we want to end up. The model that I think we should be aiming for is:
```
# dtype is a metaclass
>>> issubclass(np.dtype, type)
True

# dtypes instances are still just that - dtype instances
>>> i = np.dtype('int32')
>>> type(i)
np.dtype

# but they're also now real python types
>>> isinstance(i, type)
True

# meaning they can be used as constructors
>>> type(i(1.0)) is i
True
```

`np.int_type` shouldn't be a subclass of dtype in my mind - it should be an _instance_ of dtype.

With that in mind, I'm not sure this patch really makes any sense - I can't think of a good reason to subclass dtype right now.

Sorry for not making this remark earlier in the review.

---

Would be great if you could split off the bug fixes here so that those can make 1.16

**Comment by mattip** [comment]
> With that in mind, I'm not sure this patch really makes any sense - I can't think of a good reason to subclass dtype right now.

This is one step toward enabling proper subclassing of dtypes, needed for projects like units and categorical dtypes. The first step is allowing creation of other dtype types (not objects), while keeping the array.dtype attribute an object. The final goal is indeed to make `array.dtype` a type object. Even better would be that making an instance of the dtype would be a scalar.

While this PR is only a minor refactoring, it fixes the order of `arraydecr_new` and the API function `PyArray_DescrNew`. The internal `tp_new` should not be calling out to an API function, so the PR is valuable on its own.

**Comment by eric-wieser** [comment]
> This is one step toward enabling proper subclassing of dtypes

As I read the task document, the intent was to subclass `dtype` _instance_s, not `dtype` itself.

I cannot think of a single use case right now for subclasses of `dtype` itself.

**Comment by mattip** [comment]
Is subclassing an instance a common python idiom? I thought subclasses must inherit from a type, not an instance. Are there restrictions on methods called on a class that inherits from an instance?

**Comment by eric-wieser** [comment]
> Is subclassing an instance a common python idiom? I thought subclasses must inherit from a type

No, you're correct. My point is that subclasses don't really make sense until we make `dtype` instances `type`s.

**Comment by mattip** [comment]
well, it is the **next** step, and will not be the last step. I need the change to use `subtype` in `PyObject_New`. I guess I could redo this PR to do that with a more minimal change, but it seems unnatural to me that `arraydescr_new`, which is `type->tp_new` is calling out to an API function to actually allocate, not the other way around.

**Comment by eric-wieser** [comment]
> I need the change to use subtype in PyObject_New.

I understand this - it's that next step specifically that I think is in the wrong direction, and something we'll end up undoing. Once dtype becomes a metatype, then we'll end up reverting back to using `PyObject_New` called on `dtype`, not subclasses.

Could you maybe update your document with the proposed steps here? Perhaps I'm not seeing the bigger picture.

> but it seems unnatural to me that arraydescr_new, which is type->tp_new is calling out to an API function to actually allocate, not the other way around.

It's worth looking at the immutable types in CPython for precedence here. `PyFrozenSet_New` does not call `frozenset->tp_new` at all. Nor is `PyLongObject->tp_new` called by any of the C api constructors.

I think it's better to view `__new__` as the python-only overloaded entry point to the construction, and not something the C api should go through. Extracting a common helper function for both to go through is fine (the `descr_from_descr` function of this patch), but since CPython doesn't do it either, I don't think it makes sense for the C api to round trip through the python api.

**Comment by mattip** [comment]
> Could you maybe update your document with the proposed steps here? Perhaps I'm not seeing the bigger picture.

I am kind of working it out as I go. I think I know what I want, but I am not sure how to get there. Any help or critique is helpful, thanks for the insights.

**Comment by mattip** [comment]
Closing this in favor of other approach

**Pull request closed without merging**